modules/collect/src/main/java/com/opengamma/strata/collect/io/XmlFile.java

```diff
@@ -0,0 +1,282 @@
+  private static String parseElementName(XMLStreamReader reader) throws XMLStreamException {
+    int event = reader.getEventType();
+    while (event != XMLStreamConstants.START_ELEMENT) {
+      event = reader.next();
+    }
+    String prefix = reader.getPrefix();
+    if (prefix.length() == 0) {
+      return reader.getLocalName();
+    }
+    return prefix + ':' + reader.getLocalName();
+  }
+
+  // parses attributes into a map
+  private static ImmutableMap<String, String> parseAttributes(XMLStreamReader reader) {
+    ImmutableMap<String, String> attrs;
+    int attributeCount = reader.getAttributeCount() + reader.getNamespaceCount();
+    if (attributeCount == 0) {
+      attrs = ImmutableMap.of();
+    } else {
+      ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+      for (int i = 0; i < reader.getAttributeCount(); i++) {
+        String prefix = reader.getAttributePrefix(i);
+        if (prefix.length() == 0) {
+          builder.put(reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+        } else {
+          builder.put(prefix + ':' + reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+        }
+      }
+      for (int i = 0; i < reader.getNamespaceCount(); i++) {
+        String namespacePrefix = reader.getNamespacePrefix(i);
+        if (namespacePrefix != null) {
+          builder.put("xmlns:" + namespacePrefix, reader.getNamespaceURI(i));
+        } else {
+          builder.put("xmlns", reader.getNamespaceURI(i));
+        }
+      }
+      attrs = builder.build();
+    }
+    return attrs;
+  }
+
+  //-------------------------------------------------------------------------
+  // creates the XML input factory., recreated each time to avoid JDK-8028111.
+  private static XMLInputFactory xmlInputFactory() {
+    XMLInputFactory factory = XMLInputFactory.newFactory();
+    factory.setProperty(XMLInputFactory.IS_COALESCING, true);
+    factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, true);
+    factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false);
+    factory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+    return factory;
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Restricted constructor.
+   */
+  private XmlFile(XmlElement root, Map<String, XmlElement> refs) {
+    this.root = ArgChecker.notNull(root, "root");
+    this.refs = ImmutableMap.copyOf(ArgChecker.notNull(refs, "refs"));
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Gets the root element of this file.
+   * 
+   * @return the root element
+   */
+  public XmlElement getRoot() {
+    return root;
+  }
+
+  /**
+   * Gets the reference map of id to element.
+   * <p>
+   * This is used to decode references, such as an id/href pair.
+   * <p>
+   * For example, if one part of the XML has {@code <foo id="fooId">}, the map will
```

Comment:
Now I understand what this is all about. It would be useful to copy this example to the other places that talk about mapping references.